File: heron/api/src/java/com/twitter/heron/dsl/Resources.java
//  Copyright 2017 Twitter. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

package com.twitter.heron.dsl;

import java.io.Serializable;

/**
 * Resources needed by the topology are encapsulated in this class.
 * Currently we deal with cpu and ram. Others can be added later.
 */
public final class Resources implements Serializable {
  private static final long serialVersionUID = 630451253428388496L;
  private float cpu;
  private long ram;

  public float getCpu() {
    return cpu;
  }

  public long getRam() {
    return ram;
  }

  public Resources() {
    this.cpu = 1.0f;
    this.ram = 104857600;
  }

  public Resources withCpu(float ncpu) {
    this.cpu = ncpu;
    return this;
  }

  public Resources withRam(long nram) {
    this.ram = nram;
    return this;
  }

  public Resources withRamInMB(long nram) {
    return withRam(nram * 1024 * 1024);
  }

  public Resources withRamInGB(long nram) {
    return withRamInMB(nram * 1024);
  }

  @Override
  public String toString() {
    return "{ CPU: " + String.valueOf(cpu) + " RAM: " + String.valueOf(ram) + " }";
  }
}

File: heronpy/streamlet/resources.py
#!/usr/bin/env python3
# -*- encoding: utf-8 -*-

#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing,
#  software distributed under the License is distributed on an
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
#  specific language governing permissions and limitations
#  under the License.

'''resources.py: module for defining resources'''

class Resources:
  """Resources needed by the topology are encapsulated in this class.
     Currently we deal with CPU and RAM. Others can be added later.
  """
  def __init__(self):
    self._cpu = 0.0
    self._ram = 0

  def set_cpu(self, cpu):
    self._cpu = float(cpu)
    return self

  def set_ram(self, ram):
    self._ram = int(ram)
    return self

  def get_cpu(self):
    return self._cpu

  def get_ram(self):
    return self._ram

  def set_ram_in_mb(self, ram):
    return self.set_ram(ram * 1024 * 1024)

  def set_ram_in_gb(self, ram):
    return self.set_ram_in_mb(ram * 1024)

  def __repr__(self):
    return f'Resource {{cpu: {self._cpu}, ram: {self._ram}}}'
